react-components/src/components/RuleBasedOutputs/RuleBasedOutputs.tsx

```diff
@@ -0,0 +1,94 @@
+/*!
+ * Copyright 2023 Cognite AS
+ */
+import { useEffect, type ReactElement } from 'react';
+
+import { type IdEither, type AssetMapping3D, type CogniteClient } from '@cognite/sdk';
+import { type CogniteCadModel } from '@cognite/reveal';
+import { useReveal } from '../..';
+import { Color } from 'three';
+import { type RuleOutputSet } from './types';
+import { useSDK } from '../RevealCanvas/SDKProvider';
+import { generateRuleBasedOutputs } from './utils';
+
+export type ColorOverlayProps = {
+  ruleSet: RuleOutputSet | undefined;
+};
+
+export function RuleBasedOutputs({ ruleSet }: ColorOverlayProps): ReactElement | undefined {
+  const cdfClient = useSDK();
+
+  const viewer = useReveal();
+
+  // only enabled whether has loaded models
+  if (viewer.models === undefined || viewer.models.length === 0) return;
+
+  const models = viewer.models;
+
+  // clean up the appearance
+  models.forEach((model) => {
+    const currentModel = model as CogniteCadModel;
+    currentModel.removeAllStyledNodeCollections();
+
+    currentModel.setDefaultNodeAppearance({
+      color: new Color('#efefef')
+    });
+  });
+
+  if (ruleSet === undefined) {
+    return <></>;
+  }
+
+  useEffect(() => {
+    const getContextualization = async (model: CogniteCadModel): Promise<void> => {
+      const { modelId, revisionId } = model;
+
+      const assetMappings = await getCdfCadContextualization({
+        sdk: cdfClient,
+        modelId,
+        revisionId,
+        nodeId: undefined,
+        assetId: undefined
+      });
+
+      // remove duplicated
+      const uniqueContextualizedAssetIds = [...new Set(assetMappings.map((node) => node.assetId))];
+
+      const contextualizedAssetIds = uniqueContextualizedAssetIds.map((id) => {
+        return { id };
+      }) as unknown as IdEither[];
+
+      // get the assets with asset info from the asset ids
+      const contextualizedAssetNodes = await cdfClient.assets.retrieve(contextualizedAssetIds);
+
+      // generate rule based coloring
+      generateRuleBasedOutputs(model, contextualizedAssetNodes, assetMappings, ruleSet);
+    };
+
+    models.forEach((model) => {
+      void getContextualization(model as CogniteCadModel);
+    });
+  }, [ruleSet]);
+
+  return <></>;
+}
+
+const getCdfCadContextualization = async ({
+  sdk,
+  modelId,
+  revisionId,
+  nodeId,
+  assetId
+}: {
+  sdk: CogniteClient;
```

Comment:
We already have a way of getting this - try `useAllMappedEquipmentAssetMappings`, it is also being cached

Reply:
Ah ok cool.
Changed here:
https://github.com/cognitedata/reveal/pull/4210/commits/d748037cc48301eb04ef215bd184c86e4158a622